test/integration/infrastructure/infrastructure_test.go

```diff
@@ -0,0 +1,760 @@
+
+		internalChartsPath string
+		clientSet          *azureClientSet
+	)
+
+	BeforeSuite(func() {
+		flag.Parse()
+		validateFlags()
+
+		internalChartsPath = azure.InternalChartsPath
+		repoRoot := filepath.Join("..", "..", "..")
+		azure.InternalChartsPath = filepath.Join(repoRoot, azure.InternalChartsPath)
+		logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)))
+
+		log := logrus.New()
+		log.SetOutput(GinkgoWriter)
+		logger = logrus.NewEntry(log)
+
+		By("starting test environment")
+		testEnv = &envtest.Environment{
+			UseExistingCluster: pointer.BoolPtr(true),
+			CRDInstallOptions: envtest.CRDInstallOptions{
+				Paths: []string{
+					filepath.Join(repoRoot, "example", "20-crd-cluster.yaml"),
+					filepath.Join(repoRoot, "example", "20-crd-infrastructure.yaml"),
+				},
+			},
+		}
+
+		cfg, err := testEnv.Start()
+		Expect(err).NotTo(HaveOccurred())
+		mgr, err := manager.New(cfg, manager.Options{})
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(extensionsv1alpha1.AddToScheme(mgr.GetScheme())).To(Succeed())
+		Expect(azureinstall.AddToScheme(mgr.GetScheme())).To(Succeed())
+
+		Expect(infrastructure.AddToManager(mgr)).To(Succeed())
+
+		var mgrContext context.Context
+		mgrContext, mgrCancel = context.WithCancel(ctx)
+
+		By("start manager")
+		go func() {
+			err := mgr.Start(mgrContext.Done())
+			Expect(err).NotTo(HaveOccurred())
+		}()
+
+		c = mgr.GetClient()
+		Expect(c).NotTo(BeNil())
+		decoder = serializer.NewCodecFactory(mgr.GetScheme()).UniversalDecoder()
+
+		authorizer, err := getAuthorizer(*tenantId, *clientId, *clientSecret)
+		Expect(err).NotTo(HaveOccurred())
+		clientSet = newClientSet(*subscriptionId, authorizer)
+	})
+
+	AfterSuite(func() {
+		defer func() {
+			By("stopping manager")
+			mgrCancel()
+		}()
+
+		By("running cleanup actions")
+		framework.RunCleanupActions()
+
+		By("stopping test environment")
+		Expect(testEnv.Stop()).To(Succeed())
+
+		azure.InternalChartsPath = internalChartsPath
+	})
+
+	Context("with availabilitySet cluster creating new vNet", func() {
+		AfterEach(func() {
+			framework.RunCleanupActions()
+		})
+
+		It("should successfully create and delete", func() {
+			providerConfig := newProviderConfig(nil, false, false)
+
+			namespace, err := generateName()
+			Expect(err).NotTo(HaveOccurred())
+
+			err = runTest(ctx, logger, c, clientSet, namespace, providerConfig, decoder)
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
+
+	Context("with zonal cluster creating new vNet and NatGateway", func() {
+		AfterEach(func() {
+			framework.RunCleanupActions()
+		})
+
+		It("should successfully create and delete", func() {
+			providerConfig := newProviderConfig(nil, true, true)
+
+			namespace, err := generateName()
+			Expect(err).NotTo(HaveOccurred())
+
+			err = runTest(ctx, logger, c, clientSet, namespace, providerConfig, decoder)
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
+
+	Context("with zonal cluster using existing vNet and NatGateway", func() {
```

Comment:
Should we add also a test for existing vnet + availabilityset based cluster?

Reply:
Added one, with the addition of a "foreign" identity